src/message_expander.lua

```diff
@@ -0,0 +1,167 @@
+        local guard = transition[2]
+        if(from_state == self.state and guard(token)) then
+            local action = transition[3]
+            action(self, token)
+            local to_state = transition[4]
+            return to_state
+        end
+    end
+end
+
+local function is_any()
+    return true
+end
+
+local function is_opening_bracket(token)
+    return token == "{"
+end
+
+local function is_closing_bracket(token)
+    return token == "}"
+end
+
+local function is_pipe(token)
+
+    return token == "|"
+end
+
+local function is_u (token)
+    return token == "u"
+end
+
+local function is_not_bracket(token)
+    return not is_opening_bracket(token) and not is_closing_bracket(token)
+end
+
+local function add_token(self, token)
+    table.insert(self.tokens_, token)
+end
+
+local function add_open(self, token)
+    table.insert(self.tokens_, "{")
+    table.insert(self.tokens_, token)
+end
+
+local function add_parameter_name(self, token)
+    table.insert(self.last_parameter_.characters, token)
+end
+
+local function set_unquoted(self)
+    self.last_parameter_.quote = false
+end
+
+local function unwrap_parameter_value(parameter)
+    if parameter == nil then
+        return "missing value"
+    else
+        local type = type(parameter)
+        if (type == "table") then
+            return parameter.value
+        else
+            return parameter
+        end
+    end
+end
+
+local function replace_parameter(self)
+    local parameter_name = table.concat(self.last_parameter_.characters)
+    local value = unwrap_parameter_value(self.parameters[parameter_name])
+    local type = type(value)
+    if (type == "string") and (self.last_parameter_.quote) then
+        table.insert(self.tokens_, "'")
+        table.insert(self.tokens_, value)
+        table.insert(self.tokens_, "'")
+    else
+        table.insert(self.tokens_, value)
+    end
+    self.last_parameter_.characters = {}
+    self.last_parameter_.quote = true
+end
+
+local function replace_and_add(self, token)
+    replace_parameter(self)
+    add_token(self, token)
+end
+
+local function do_nothing() end
+
+M.transitions_ = {
+    {"TEXT"              , is_not_bracket    , add_token         , "TEXT"              },
+    {"TEXT"              , is_opening_bracket, do_nothing        , "OPEN_1"            },
+    {"OPEN_1"            , is_opening_bracket, do_nothing        , "PARAMETER"         },
+    {"OPEN_1"            , is_any            , add_open          , "TEXT"              },
+    {"PARAMETER"         , is_closing_bracket, do_nothing        , "CLOSE_1"           },
+    {"PARAMETER"         , is_pipe           , do_nothing        , "SWITCH"            },
+    {"PARAMETER"         , is_any            , add_parameter_name, "PARAMETER"         },
+    {"SWITCH"            , is_closing_bracket, do_nothing        , "CLOSE_1"           },
+    {"SWITCH"            , is_u              , set_unquoted      , "SWITCH"            },
```

Comment:
While `uq` is possible (basically by ignoring the `q`), `u` as a switch is better. Because if you want to add more switches, they all should be only one character wide.

Reply:
Also for fault tolerance I simply ignore switches that I don't know.

I was thinking about adding a validation method, but that would be pointless, because when you need to produce an error, validation does not help anymore. It is too late for that. You want to safe as much information as possible.